<a href="https://colab.research.google.com/github/PawanKumar1216/northstar-databases-analytics-assignment/blob/main/03_Dataset_Overview.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 03 Dataset Overview

## NorthStar Urban Mobility and Logistics

This notebook imports and inspects the NorthStar operational datasets. It provides an overview of the files, dataset dimensions, missing values, and relationships between the main data sources used in the project.

## 1. Import Required Libraries and Load Dataset Files from Github

The original NorthStar dataset contains multiple connected files covering deliveries, orders, hubs, customers, complaints, incidents, drivers, vehicles, and app events. These files are loaded into Python using the Pandas library so that the overall dataset structure can be inspected before later SQL, R, Python, and MongoDB analysis.

## 2. Dataset Dimensions and Structure

The first stage of inspection is to check the size of the main NorthStar datasets. This shows how many records and variables are available across the connected operational files.

In [2]:
# Check the dimensions of the key operational datasets
dataset_dimensions = pd.DataFrame({
    "Dataset": [
        "Customers",
        "Deliveries",
        "Orders",
        "Complaints",
        "App Events",
        "Drivers",
        "Hubs",
        "Incidents",
        "Vehicles"
    ],
    "Records": [
        customers.shape[0],
        deliveries.shape[0],
        orders.shape[0],
        complaints.shape[0],
        app_events.shape[0],
        drivers.shape[0],
        hubs.shape[0],
        incidents.shape[0],
        vehicles.shape[0]
    ],
    "Columns": [
        customers.shape[1],
        deliveries.shape[1],
        orders.shape[1],
        complaints.shape[1],
        app_events.shape[1],
        drivers.shape[1],
        hubs.shape[1],
        incidents.shape[1],
        vehicles.shape[1]
    ]
})

dataset_dimensions

,Dataset,Records,Columns
0,Customers,650,9
1,Deliveries,950,13
2,Orders,1250,11
3,Complaints,320,10
4,App Events,640,10
5,Drivers,170,8
6,Hubs,8,5
7,Incidents,280,7
8,Vehicles,120,8


In [1]:
import pandas as pd

base_url = "https://raw.githubusercontent.com/PawanKumar1216/northstar-databases-analytics-assignment/main/data/"

deliveries = pd.read_csv(base_url + "deliveries.csv")
orders = pd.read_csv(base_url + "orders.csv")
hubs = pd.read_csv(base_url + "hubs.csv")
customers = pd.read_csv(base_url + "customers.csv")
app_events = pd.read_csv(base_url + "app_events.csv")
complaints = pd.read_csv(base_url + "complaints.csv")
incidents = pd.read_csv(base_url + "incidents.csv")
vehicles = pd.read_csv(base_url + "vehicles.csv")
drivers = pd.read_csv(base_url + "drivers.csv")
data_dictionary = pd.read_csv(base_url + "data_dictionary.csv")

print("All NorthStar dataset files loaded successfully from GitHub.")

All NorthStar dataset files loaded successfully from GitHub.


## 3. Missing Values and Data Quality Inspection

Missing values are checked across the main NorthStar datasets to identify incomplete operational records. This is important because missing customer, delivery, or order information can reduce data quality and affect later analysis.

In [3]:
# Check missing values in the main datasets
print("Missing Values in Customers Dataset")
print(customers.isnull().sum())

print("\nMissing Values in Deliveries Dataset")
print(deliveries.isnull().sum())

print("\nMissing Values in Orders Dataset")
print(orders.isnull().sum())

Missing Values in Customers Dataset
customer_id              0
age                      0
home_zone                0
customer_type            0
signup_date              0
loyalty_score           20
app_engagement_score     0
preferred_channel       13
account_status           0
dtype: int64

Missing Values in Deliveries Dataset
delivery_id                       0
order_id                          0
driver_id                         0
vehicle_id                        0
hub_id                            0
dispatch_time                     0
delivery_completed_at            19
delivery_status                   0
route_distance_km                 0
manual_route_override_count       0
proof_of_completion_missing       0
customer_rating_post_delivery    14
fuel_or_charge_cost               0
dtype: int64

Missing Values in Orders Dataset
order_id                  0
customer_id               0
service_type              0
order_created_at          0
promised_window_hours     0
pickup_zone    

## 4. Dataset File Preview

A small preview of the main datasets is displayed to inspect the column names, data structure, and example records before deeper analysis.

In [5]:
print("Customers Dataset Preview")
display(customers.head())

print("\nDeliveries Dataset Preview")
display(deliveries.head())

print("\nOrders Dataset Preview")
display(orders.head())

Customers Dataset Preview


,customer_id,age,home_zone,customer_type,signup_date,loyalty_score,app_engagement_score,preferred_channel,account_status
0,C0001,26,North,SME,2024-11-27 04:25:00,44.9,69.2,App,Active
1,C0002,61,AIRPORT,Consumer,2025-10-28 01:04:00,55.4,66.6,App,Active
2,C0003,66,East,Consumer,2025-07-02 03:23:00,75.9,33.8,NaN,Active
3,C0004,75,CENTRAL,Consumer,2025-08-19 01:58:00,32.5,33.0,App,Active
4,C0005,26,Riverside,Consumer,2025-06-03 06:02:00,55.9,100.0,Web,Active



Deliveries Dataset Preview


,delivery_id,order_id,driver_id,vehicle_id,hub_id,dispatch_time,delivery_completed_at,delivery_status,route_distance_km,manual_route_override_count,proof_of_completion_missing,customer_rating_post_delivery,fuel_or_charge_cost
0,DL00001,O00938,D004,V056,H05,2024-06-18 10:57:00,2024-06-19 09:05:59.904311,Failed,17.26,1,0,3.07,12.05
1,DL00002,O00004,D138,V007,H02,2025-01-11 18:45:00,2025-01-11 17:39:00.000000,OnTime,10.34,1,0,5.00,13.41
2,DL00003,O00639,D006,V049,H02,2025-06-02 20:39:00,2025-06-02 21:45:32.366770,OnTime,7.92,0,0,4.98,8.51
3,DL00004,O00313,D116,V055,H02,2024-03-08 23:31:00,2024-03-09 23:30:08.103702,Delayed,16.42,0,0,4.18,13.62
4,DL00005,O00844,D108,V034,H01,2025-09-21 11:43:00,2025-09-21 15:45:34.131056,OnTime,14.52,1,0,4.18,9.22



Orders Dataset Preview


,order_id,customer_id,service_type,order_created_at,promised_window_hours,pickup_zone,dropoff_zone,priority_level,order_value,booking_channel,special_handling_flag
0,O00001,C0292,Passenger,2024-08-20 14:43:00,6,Airport,South,Medium,126.65,App,0
1,O00002,C0459,Passenger,2024-05-14 22:16:00,24,North,AIRPORT,Low,109.30,App,0
2,O00003,C0161,Passenger,2025-09-02 14:37:00,4,West,AIRPORT,High,33.50,Phone,0
3,O00004,C0520,Parcel,2025-01-11 17:15:00,2,RiverSide,North,Medium,10.04,App,1
4,O00005,C0558,Retail,2025-02-17 19:32:00,12,Riverside,SOUTH,Low,125.58,Phone,0


## 5. Dataset Column Summary

The column names of each dataset are reviewed to understand the types of operational information stored across the NorthStar environment. This helps identify how the files can later be connected for SQL, R, Python, and MongoDB analysis.

In [6]:
# Display the column names for each key dataset
datasets = {
    "Customers": customers,
    "Deliveries": deliveries,
    "Orders": orders,
    "Complaints": complaints,
    "App Events": app_events,
    "Drivers": drivers,
    "Hubs": hubs,
    "Incidents": incidents,
    "Vehicles": vehicles
}

for name, df in datasets.items():
    print(f"{name} Columns:")
    print(list(df.columns))
    print()

Customers Columns:
['customer_id', 'age', 'home_zone', 'customer_type', 'signup_date', 'loyalty_score', 'app_engagement_score', 'preferred_channel', 'account_status']

Deliveries Columns:
['delivery_id', 'order_id', 'driver_id', 'vehicle_id', 'hub_id', 'dispatch_time', 'delivery_completed_at', 'delivery_status', 'route_distance_km', 'manual_route_override_count', 'proof_of_completion_missing', 'customer_rating_post_delivery', 'fuel_or_charge_cost']

Orders Columns:
['order_id', 'customer_id', 'service_type', 'order_created_at', 'promised_window_hours', 'pickup_zone', 'dropoff_zone', 'priority_level', 'order_value', 'booking_channel', 'special_handling_flag']

Complaints Columns:
['complaint_id', 'customer_id', 'order_id', 'complaint_type', 'channel', 'severity', 'created_at', 'status', 'resolution_days', 'compensation_amount']

App Events Columns:
['event_id', 'customer_id', 'order_id', 'event_timestamp', 'event_type', 'session_id', 'device_type', 'zone_context', 'api_latency_ms', 'suc

## 6. Cross-File Relationships

The NorthStar datasets are connected through shared identifiers. These relationships allow operational records from different business functions to be integrated for later SQL analysis, customer complaint investigation, delivery performance analysis, and MongoDB document modelling.

The main relationships are:

- `customers.customer_id` → `orders.customer_id`
- `customers.customer_id` → `complaints.customer_id`
- `customers.customer_id` → `app_events.customer_id`
- `orders.order_id` → `deliveries.order_id`
- `orders.order_id` → `complaints.order_id`
- `orders.order_id` → `app_events.order_id`
- `deliveries.delivery_id` → `incidents.delivery_id`
- `deliveries.driver_id` → `drivers.driver_id`
- `deliveries.vehicle_id` → `vehicles.vehicle_id`
- `deliveries.hub_id` → `hubs.hub_id`

In [7]:
# Verify key relationships between the main NorthStar datasets
relationship_check = pd.DataFrame({
    "Relationship": [
        "Orders linked to Customers",
        "Complaints linked to Customers",
        "App Events linked to Customers",
        "Deliveries linked to Orders",
        "Complaints linked to Orders",
        "App Events linked to Orders",
        "Incidents linked to Deliveries",
        "Deliveries linked to Drivers",
        "Deliveries linked to Vehicles",
        "Deliveries linked to Hubs"
    ],
    "Matched Records": [
        orders["customer_id"].isin(customers["customer_id"]).sum(),
        complaints["customer_id"].isin(customers["customer_id"]).sum(),
        app_events["customer_id"].isin(customers["customer_id"]).sum(),
        deliveries["order_id"].isin(orders["order_id"]).sum(),
        complaints["order_id"].isin(orders["order_id"]).sum(),
        app_events["order_id"].isin(orders["order_id"]).sum(),
        incidents["delivery_id"].isin(deliveries["delivery_id"]).sum(),
        deliveries["driver_id"].isin(drivers["driver_id"]).sum(),
        deliveries["vehicle_id"].isin(vehicles["vehicle_id"]).sum(),
        deliveries["hub_id"].isin(hubs["hub_id"]).sum()
    ],
    "Total Records": [
        len(orders),
        len(complaints),
        len(app_events),
        len(deliveries),
        len(complaints),
        len(app_events),
        len(incidents),
        len(deliveries),
        len(deliveries),
        len(deliveries)
    ]
})

relationship_check

,Relationship,Matched Records,Total Records
0,Orders linked to Customers,1250,1250
1,Complaints linked to Customers,320,320
2,App Events linked to Customers,640,640
3,Deliveries linked to Orders,950,950
4,Complaints linked to Orders,320,320
5,App Events linked to Orders,496,640
6,Incidents linked to Deliveries,280,280
7,Deliveries linked to Drivers,950,950
8,Deliveries linked to Vehicles,950,950
9,Deliveries linked to Hubs,950,950


## 7. Initial Delivery Status Inspection

Delivery status is inspected to provide an early view of service reliability within the NorthStar operational dataset. The distribution of on-time, delayed, and failed deliveries helps identify whether the organisation is experiencing meaningful delivery-performance issues.

In [8]:
# Count delivery records by delivery status
delivery_status_summary = deliveries["delivery_status"].value_counts().reset_index()
delivery_status_summary.columns = ["Delivery Status", "Number of Deliveries"]

delivery_status_summary

,Delivery Status,Number of Deliveries
0,OnTime,616
1,Delayed,202
2,Failed,132


## 8. Summary of Dataset Overview

The dataset overview confirmed that the NorthStar project contains multiple connected operational files covering customers, orders, deliveries, complaints, drivers, hubs, vehicles, incidents, and app events. The dataset includes both structured relational records and flexible event-based data, making it suitable for combined SQL, R, Python, and MongoDB analysis.

The initial inspection identified several important data-quality issues, including missing values in customer loyalty scores, preferred channels, delivery completion timestamps, customer delivery ratings, and order booking channels. Cross-file relationship checks also showed that most operational records are fully linked, although some app-event records are not connected to order records, reflecting the fragmented and semi-structured data environment described in the case study.

The delivery-status inspection identified 616 on-time deliveries, 202 delayed deliveries, and 132 failed deliveries. This means that approximately 35.2% of delivery records were delayed or failed, providing early evidence of service-reliability issues that require deeper investigation in later SQL, R, and Python analysis.